components/lib/crash/src/test/java/mozilla/components/lib/crash/service/GleanCrashReporterServiceTest.kt

```diff
@@ -0,0 +1,112 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.lib.crash.service
+
+import android.content.Context
+import androidx.test.core.app.ApplicationProvider
+import androidx.work.testing.WorkManagerTestInitHelper
+import mozilla.components.lib.crash.Crash
+import mozilla.components.lib.crash.GleanMetrics.CrashMetrics
+import mozilla.components.service.glean.Glean
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertTrue
+import org.junit.Assert.assertFalse
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.mockito.Mockito.spy
+import org.mockito.Mockito.verify
+import org.mockito.Mockito.verifyNoMoreInteractions
+import org.robolectric.RobolectricTestRunner
+
+@RunWith(RobolectricTestRunner::class)
+class GleanCrashReporterServiceTest {
+    private val context: Context
+        get() = ApplicationProvider.getApplicationContext()
+
+    private fun setupGleanCrashReporterServiceTest() {
+        WorkManagerTestInitHelper.initializeTestWorkManager(
+            ApplicationProvider.getApplicationContext())
+
+        Glean.enableTestingMode()
+        Glean.initialize(context)
+    }
+
+    @Test
+    fun `GleanCrashReporterService records native code crashes`() {
```

Comment:
If we do the refactoring above, we could do real unit testing, in isolation. We could test that:

* the function that writes crashes:
  - correctly writes from the expected processes;
  - works if the file doesn't exist yet;
  - doesn't crash if it can't write to the file;
  - the data is written in the expected format.

* the function that reads crashes:
  - only works from the main process;
  - correctly parses the data;
  - doesn't crash if the data is corrupted;
  - properly asserts that Glean is initialized.

Reply:
I'm okay with adding the additional test coverage, the only problem being with testing around processes.  As far as I know, it's not possible to test multiple processes for background things like this.  I did find [Multiprocess Espresso](https://developer.android.com/training/testing/espresso/multiprocess) but it looks like it's purpose is for testing UI interactions across multiple processes.

Reply:
@travis79 I think the tests above are still missing. I think it's ok to not have the process specific tests, but the others that I called out should really be there.

Reply:
> If we do the refactoring above, we could do real unit testing, in isolation. We could test that:
> 
>     * the function that writes crashes:
>       
>       * correctly writes from the expected processes;
I believe that this is covered by the first 3 tests, checking that it records and can read back the 2 types of crashes as well as being able to process multiple crashes from a single file
>       * works if the file doesn't exist yet;
Each of the first 3 tests check to see that no file exists before starting.
>       * doesn't crash if it can't write to the file;
I added a test for this
>       * the data is written in the expected format.
Each of the first 3 tests check for content of the file.
> 
>     * the function that reads crashes:
>       
>       * only works from the main process;
No way to easily test multi-process stuff in unit tests, can file a bug for follow up if you agree
>       * correctly parses the data;
The first 3 tests check this by verifying data through the Glean TestAPI
>       * doesn't crash if the data is corrupted;
I added a test for this
>       * properly asserts that Glean is initialized.
I added a log message to indicate when the lib-crash interaction that requires the main process happens off of the main process.